Extensions/PopUp.json

```diff
@@ -0,0 +1,282 @@
+      "events": [
+        {
+          "disabled": false,
+          "folded": false,
+          "type": "BuiltinCommonInstructions::JsCode",
+          "inlineCode": "const id = eventsFunctionContext.getArgument(\"ID\");\nconst message = eventsFunctionContext.getArgument(\"Message\");\nconst defaultText = eventsFunctionContext.getArgument(\"DefaultText\");\n\nconst response = window.prompt(message, defaultText);\n\nif (response !== null){\n    runtimeScene.getGame().getVariables().get(id).setString(response);\n}else{\n    runtimeScene.getGame().getVariables().get(id).setBoolean(false);\n}",
+          "parameterObjects": "",
+          "useStrict": true,
+          "eventsSheetExpanded": true
+        }
+      ],
+      "parameters": [
+        {
+          "codeOnly": false,
+          "defaultValue": "",
+          "description": "Prompt message",
+          "longDescription": "",
+          "name": "Message",
+          "optional": false,
+          "supplementaryInformation": "",
+          "type": "string"
+        },
+        {
+          "codeOnly": false,
+          "defaultValue": "",
+          "description": "Input placeholder",
+          "longDescription": "",
+          "name": "DefaultText",
+          "optional": false,
+          "supplementaryInformation": "",
+          "type": "string"
+        },
+        {
+          "codeOnly": false,
+          "defaultValue": "",
+          "description": "Pop-up identifier",
+          "longDescription": "",
+          "name": "ID",
+          "optional": false,
+          "supplementaryInformation": "",
+          "type": "string"
+        }
+      ],
+      "objectGroups": []
+    },
+    {
+      "description": "Ask confirmation of user with a message in a dialog box with an OK button, and a Cancel button.",
+      "fullName": "Confirm",
+      "functionType": "Action",
+      "group": "",
+      "name": "Confirm",
+      "private": false,
+      "sentence": "Open a confirmation pop-up box with message: _PARAM1__ (identifier: _PARAM2_)",
+      "events": [
+        {
+          "disabled": false,
+          "folded": false,
+          "type": "BuiltinCommonInstructions::Standard",
+          "conditions": [
+            {
+              "type": {
+                "inverted": false,
+                "value": "BuiltinCommonInstructions::Once"
+              },
+              "parameters": [],
+              "subInstructions": []
+            }
+          ],
+          "actions": [],
+          "events": [
+            {
+              "disabled": false,
+              "folded": false,
+              "type": "BuiltinCommonInstructions::JsCode",
+              "inlineCode": "const id = eventsFunctionContext.getArgument(\"ID\");\nconst isConfirmed = confirm(eventsFunctionContext.getArgument(\"Message\"));\n\nruntimeScene.getGame().getVariables().get(id).setBoolean(isConfirmed);\n\n",
```

Comment:
This is an action that opens the confirmation box.
Not sure to understand your opinion.

Here mine:
I think it should be an action because the user wants **to do** something, he asks the game engine to opens a confirmation box, then the response is saved to be checked in another extension condition.
If the user search for open a confirmation box he'll never search for it in the condition list.

Reply:
To me an action do something or ask for something and a condition check if the actions or expressions are valid or not.

Opening a popup box is an action.
Check the returned value is a condition.

Never a condition should do or trigger an action.

Reply:
Fair point, though please don't use the id as a global variable name, use a structure instead, as the best practices dictate. I'd also use a single scene variable without an ID, since those expressions are meant to get the result directly afterwards, otherwise, all inputted data is leaked. If a user wants to keep the data longer or outside the scene, well, variables exists for that purpose.

Reply:
I've rewrote the extension, but yes I agree that the variable structure should be used, I ask for it often and haven't done it here lol.
Thanks for your comment!

Reply:
Sorry I was gone for a bit and couldn't work on this Thanks @bouh for the changes. They made the extension much better and @ Thanks @arthuro555 for the review! I am working on it!

I agree making the variables global is a teriible idea. Could it be made clear how to init and use a structure in js? I am having some issues 😅

> I'd also use a single scene variable without an ID, since those expressions are meant to get the result directly afterwards, otherwise, all inputted data is leaked. If a user wants to keep the data longer or outside the scene, well, variables exists for that purpose.

It's just convenient to have the response persist and accessable. For instance for a login system. The user's response might be called multiple times, and it's convenient that the user original response is always available. I wanted it work similar to the tweens where you use the ID to access the tweens.

:)

Reply:
Yep no worries, I started to fix the last point of the review.
Should be done for this Friday :)

Reply:
> It's just convenient to have the response persist and accessable. For instance for a login system. The user's response might be called multiple times, and it's convenient that the user original response is always available.

Again, variables are here for that purpose. Just like a text input object's text cannot be accessed after it was deleted, after the popup is closed, the input should be given to the user and then discarded. Wanting to have a value persisted is a valid thing, but it can be easily done by storing that value in a variable. Deleting stored values is not easily doable without reading the extension source code and messing with its internal variables, which most users will never ever think of doing.

Keeping all the values stored will cause a memory leak, which most users may not want, especially if they use it for example for inputting a password: anyone that has access to the opened game can easily with the JavaScript console access the variables and read the cleartext password as that temporary value was never cleared out of memory. Another example would be loading a custom map, if you have a huge JSON describing a custom map that you paste into prompt, then this house string will occupy computer resources for nothing since after parsing it, the user no longer needs the JSON text input.

> I wanted it work similar to the tweens where you use the ID to access the tweens.

The tweens' IDs are very different, since:

1. They are needed to interact with the different currently running tweens, for example, pausing or resuming one. The popups do not have multiple long-running processes that need differentiation, there is always only one that instantly yields a result.
2. Tweens are bound to an object's lifetime, so we know that once the object is deleted, they will be deleted too. A popup's result value will never be deleted since they are bound to the game's lifetime.

Reply:
> Or just explain how to use the extension in the description?

Since the description is markdown, we can also provide code snippets in the form of screenshots to clarify the usage. I have done this with the advanced P2P event handling extension for example.